### PR TITLE
DimensionSelector-related bug fixes and optimizations (fixes #3799, part of #3798)

### LINKED ISSUE #3799: Investigate IncrementalIndex ValueMatcher performance
@leventov raised a concern in review of #3797 that even though IncrementalIndex ValueMatcher benchmarks look fine, this may not hold up in production when different strategy types are mixed. We should endeavor to write a better benchmark that smokes out any potential performance regressions, and if regressions are found, fix them before release of 0.10.0. I think FilteredAggregatorFactory benchmarks are a good place to try to verify this.

See review comment in https://github.com/druid-io/druid/pull/3797#discussion_r93735214 and response in https://github.com/druid-io/druid/pull/3797#discussion_r93792025.

### PR DESCRIPTION
**Pull request by leventov**
Fixes #3799, part of #3798

 * Add `DimensionSelector.idLookup()` and `nameLookupPossibleInAdvance()` to allow better inspection of features `DimensionSelector`s supports, it enables safer/more correct code working with `DimensionSelector`s in `BaseTopNAlgorithm`, `BaseFilteredDimensionSpec`, `DimensionSelectorUtils`.
 * Add `PredicateFilteringDimensionSelector`, to make `BaseFilteredDimensionSpec` to be able to decorate `DimensionSelector`s with unknown cardinality.
 * Add `DimensionSelector.makeValueMatcher()` (two kinds) for `DimensionSelector`-side specifics-aware optimizations of `ValueMatcher`s. Somewhere optimization goes beyond the level we had prior to #3797.
 * Optimize `getRow()` in `BaseFilteredDimensionSpec`'s `DimensionSelector` (which is refactored as `ForwardingFilteredDimensionSelector`), `StringDimensionIndexer`'s `DimensionSelector` and `SingleScanTimeDimSelector`.
 * Use two singletons, `TrueValueMatcher` and `FalseValueMatcher`, instead of `BooleanValueMatcher`.
 * Add `NullStringObjectColumnSelector` singleton and use it in `MapVirtualColumn`

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/DimensionSelector.java [hunk 1]
```diff
@@ -80,10 +90,19 @@
   public String lookupName(int id);
 
   /**
-   * The ID is the int id value of the field.
+   * Returns true if it is possible to {@link #lookupName(int)} by ids from 0 to {@link #getValueCardinality()}
+   * Before the rows with those ids are returns.
    *
-   * @param name field name to look up the id for
-   * @return the id for the given field name
+   * <p>Returns false if {@link #lookupName(int)} could be called only with ids, returned from the last call of {@link
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: suggest rewording "last call" and "last IndexedInts" to "most recent call" and "most recent IndexedInts"

File: processing/src/main/java/io/druid/query/dimension/ForwardingFilteredDimensionSelector.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+      return this;
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public int lookupId(String name)
+  {
+    // Don't cache selector.idLookup(), because it is likely implemented as `return this`, in this case caching only
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/dimension/ForwardingFilteredDimensionSelector.java]: For my own learning, what's the JVM behavior that results in the caching being slower when idLookup() returns `this`?

File: processing/src/main/java/io/druid/segment/DimensionSelector.java [hunk 3]
```diff
@@ -80,10 +90,19 @@
   public String lookupName(int id);
 
   /**
-   * The ID is the int id value of the field.
+   * Returns true if it is possible to {@link #lookupName(int)} by ids from 0 to {@link #getValueCardinality()}
+   * Before the rows with those ids are returns.
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: returns -> returned

File: processing/src/main/java/io/druid/query/dimension/ForwardingFilteredDimensionSelector.java (hunk 2 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/dimension/ForwardingFilteredDimensionSelector.java]: My idea was that if `lookupId()` is called in a hot loop along with e. g. `getRows()`, `selector` field is read once, but if we cache selector's IdLookup in another field (which will appear to point to the same selector object, but runtime won't know about this), we will read two fields.

But I missed that `lookupId()` is never called in a hot loop, so I changed to caching for simplicity.

File: processing/src/main/java/io/druid/segment/DimensionSelector.java (hunk 1 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: Fixed

File: processing/src/main/java/io/druid/segment/DimensionSelector.java (hunk 3 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: Fixed

**Comment by leventov** [comment]
@jon-wei thanks for so quick review. Addressed comments.

**Comment by gianm** [comment]
Taking a look now.

File: processing/src/main/java/io/druid/segment/DimensionSelector.java [hunk 4]
```diff
@@ -34,6 +40,10 @@
    */
   public IndexedInts getRow();
 
+  ValueMatcher makeValueMatcher(String value, boolean matchNull);
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: Fixed

File: processing/src/main/java/io/druid/segment/DimensionSelector.java [hunk 5]
```diff
@@ -34,6 +40,10 @@
...
+
+  ValueMatcher makeValueMatcher(Predicate<String> predicate, boolean matchNull);
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: Fixed

File: processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+import java.util.Objects;
+
+public final class DimensionSelectorUtils
+{
+
+  private DimensionSelectorUtils()
+  {
+  }
+
+  public static ValueMatcher makeRowBasedValueMatcher(
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java]: I wanted to emphasize the difference between DimensionSelector.makeValueMatcher() (interface methods) and DimensionSelectorUtils.makeValueMatcher() ("default" implementations of that interface methods, to which interface implementations may delegate).

Renamed to "makeValueMatcherGeneric" and added some Javadoc

File: processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+      DimensionSelector selector,
+      String value,
+      boolean matchNull
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java]: Fixed

File: processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+  )
+  {
+    IdLookup idLookup = selector.idLookup();
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java]: IMO it doens't make sense because the "entry point" to `DimensionSelector.makeValueMatcher()` is `StringValueMatcherColumnSelectorStrategy`. This optimization is applied here. And `DimensionSelectorUtils`'s methods are called _from_ `DimensionSelector` implementations.

Also I removed `(cardinality == 1 && selector.lookupName(0) == null)` part in `StringValueMatcherColumnSelectorStrategy`, because it really means "this DimensionSelector is `NullDimensionSelector`", so delegate to `NullDimensionSelector` which also returns only `BooleanValueMatcher`.

File: processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    if (idLookup != null) {
+      return makeDictionaryEncodedRowBasedValueMatcher(selector, idLookup.lookupId(value), matchNull);
+    } else if (selector.getValueCardinality() >= 0 && selector.nameLookupPossibleInAdvance()) {
+      // Employ precomputed BitSet optimization
+      return makeRowBasedValueMatcher(selector, Predicates.equalTo(value), matchNull);
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java]: Fixed

File: processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+          return false;
+        }
+      }
+    };
+  }
+
+  public static ValueMatcher makeRowBasedValueMatcher(
+      final DimensionSelector selector,
+      final Predicate<String> predicate,
+      final boolean matchNull
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java]: Fixed

File: processing/src/main/java/io/druid/segment/DimensionSelector.java [hunk 11]
```diff
@@ -80,10 +90,19 @@
...
+   * before the rows with those ids are returned.
    *
-   * @param name field name to look up the id for
-   * @return the id for the given field name
+   * <p>Returns false if {@link #lookupName(int)} could be called only with ids, returned from the most recent call of
+   * {@link #getRow()} on this DimensionSelector, but not earlier and not later. If {@link #lookupName(int)} is called
+   * with an id, which is not present in the most recent {@link IndexedInts}, returned from {@link #getRow()}, result is
+   * undefined: exception could be thrown, or null returned, or some other random value.
+   */
+  boolean nameLookupPossibleInAdvance();
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: I think such change is a subject for a separate PR, which adds support of topN processing with DimensionSelectors without cardinality, and probably also removes the cache from `SingleScanTimeDimSelector`

Comment (gianm) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: That's fair. Could you at least fix the javadoc? The "not earlier and not later" part isn't true for SingleScanTimeDimSelector (it does allow "later").

File: processing/src/main/java/io/druid/query/topn/BaseTopNAlgorithm.java [hunk 12]
```diff
@@ -232,8 +235,8 @@ public void keepOnlyN(int n)
     {
       int startIndex = ignoreFirstN;
 
-      if (previousStop != null) {
-        int lookupId = dimSelector.lookupId(previousStop) + 1;
+      if (previousStop != null && idLookup != null) {
+        int lookupId = idLookup.lookupId(previousStop) + 1;
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/topn/BaseTopNAlgorithm.java]: Reverted to the former semantics, throwing `UnsupportedOperationException`

File: processing/src/main/java/io/druid/segment/DimensionSelector.java (hunk 11 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: Changed Javadoc so it covers the special case "nameLookupAvailableInAdvance() returns false, but getValueCardinality() returns a non-negative number", only `SingleScanTimeDimSelector` has this combination.

**Comment by leventov** [comment]
Thanks @gianm, addressed comments.

**Comment by jon-wei** [comment]
Reviewed the latest updates since my previous comment, still LGTM
